### PR TITLE
perf(test): Speed up the Vitest unit-test suite

### PR DESCRIPTION
**Pull request by prakharsingh-74**
fixes - #3996 

## Which problem is this PR solving?
- Resolves the issue where the Vitest unit-test suite has a slow execution duration due to excessive module compilation overhead and redundant component rendering.

## Description of the changes
- **Vitest Config Optimization**: Narrowed down `server.deps.inline` in `vitest.config.ts` from `true` to specifically target `['react-router', 'cookie']`. This avoids forcing Vitest to compile the entire `node_modules` directory.
- **TraceStatistics Test Consolidation**: Consolidated 13 separate test cases in `TraceStatistics/index.test.jsx` into a single test case to prevent mounting/rendering the heavy component multiple times.
- **OperationDetailsTable Test Consolidation**: Consolidated 11 test cases under `<OperationTableDetails> with data` in `operationDetailsTable/index.test.jsx` into a single unified test case.

## How was this change tested?
- Verified test suite correctness and execution speed of the modified test files:
  - `TraceStatistics/index.test.jsx` duration reduced from ~7.2s to ~4.15s.
  - `operationDetailsTable/index.test.jsx` duration reduced from ~5.3s to ~4.67s.
- Ran the full test suite (`npm test`) to ensure all 251 test files and 3,058 assertions pass successfully.

image

image

## Checklist
- [x] I have read https://github.com/jaegertracing/jaeger/blob/main/CONTRIBUTING_GUIDELINES.md
- [x] I have signed all commits
- [x] I have added unit tests for the new functionality
- [x] I have run lint and test steps successfully: `make lint test`

## AI Usage in this PR (choose one)
See [AI Usage Policy](https://github.com/jaegertracing/jaeger/blob/main/CONTRIBUTING_GUIDELINES.md#ai-usage-policy).
- [] **None**: No AI tools were used in creating this PR
- [x] **Light**: AI provided minor assistance (formatting, simple suggestions)
- [] **Moderate**: AI helped with code generation or debugging specific parts
- [] **Heavy**: AI generated most or all of the code changes

### CODE REVIEW

File: packages/jaeger-ui/vitest.config.ts [hunk 1]
```diff
@@ -50,12 +50,7 @@ export default defineConfig({
...
   server: {
     deps: {
-      // Inline all node_modules through Vite's transform pipeline.
-      // The threads pool uses SSR resolve conditions that prefer .mjs files,
-      // causing CJS/ESM named-export interop failures for packages like
-      // react-router and cookie. inline:true fixes this class of problem
-      // universally rather than playing whack-a-mole with individual packages.
-      inline: true,
+      inline: ['react-router', 'cookie'],
     },
```
[the remark this answers is not in the document]
Comment (prakharsingh-74) [reply packages/jaeger-ui/vitest.config.ts:57]: Using inline: true forces Vitest to compile and transform all of node_modules on every run, which was causing the test suite to spend most of its time in infrastructure overhead (compilation, imports, environment setup). Restricting it to ['react-router', 'cookie'] solves the CJS/ESM interop issues for the affected packages while letting the rest resolve natively, vastly speeding up the test startup.

File: packages/jaeger-ui/src/components/TracePage/TraceStatistics/index.test.jsx [hunk 2]
```diff
@@ -22,23 +22,8 @@ describe('<TraceTagOverview>', () => {
...
-    expect(screen.getByRole('table')).toBeInTheDocument();
-    expect(screen.queryByText(/Tag: "SQL"/)).not.toBeInTheDocument();
-  });
-
-  it('check search', async () => {
-    const searchSet = new Set();
-    searchSet.add('service1	op1	__LEAF__');
-
+  it('does not explode, renders trace statistics, and handles search, interactive options, sorting, and searchInTable', async () => {
+    // 1. Initial render
```
[the remark this answers is not in the document]
Comment (prakharsingh-74) [reply packages/jaeger-ui/src/components/TracePage/TraceStatistics/index.test.jsx]: This consolidation was done to resolve the performance bottleneck identified by the mentor, where mounting and rendering the heavy <TraceStatistics> component 13 times was a major contributor to the test suite's CPU overhead.

While this does trade off some failure isolation, the steps here are fully synchronous and deterministic (no network or asynchronous flakiness), meaning they are highly stable.

As a compromise to maintain performance gains while improving isolation, we could separate the helper functions (like searchInTable and pure utilities, which do not need to mount the component at all) from the rendering/interaction tests. Let me know if you'd prefer that split!

File: packages/jaeger-ui/src/components/Monitor/ServicesView/operationDetailsTable/index.test.jsx [hunk 3]
```diff
@@ -74,151 +74,85 @@ describe('<OperationTableDetails> with data', () => {
...
-    const cloneServiceOpsMetrics = { ...serviceOpsMetrics[0] };
-    cloneServiceOpsMetrics.latency = 0.2988;
-    const { container } = render(
-      <OperationTableDetails {...props} data={[cloneServiceOpsMetrics]} loading={false} />
-    );
-    expect(container.querySelector('table')).toBeInTheDocument();
-  });
+  it('renders data, formats values correctly, supports sorting, row hover highlights, and tracks events', () => {
+    const trackSortOperationsSpy = jest.spyOn(track, 'trackSortOperations');
+    const trackViewTracesSpy = jest.spyOn(track, 'trackViewTraces');
```
[the remark this answers is not in the document]
Comment (prakharsingh-74) [reply packages/jaeger-ui/src/components/Monitor/ServicesView/operationDetailsTable/index.test.jsx]: fixed

File: packages/jaeger-ui/src/components/TracePage/TraceStatistics/index.test.jsx [hunk 4]
```diff
@@ -170,19 +134,17 @@ describe('<TraceTagOverview>', () => {
...
+      componentInstance.setState({
+        ...componentInstance.state,
+        tableValue: [
+          makeRow('parent-a', false, 'none'),
+          makeRow('detail-a1', true, 'parent-a', { hasSubgroupValue: false }),
+          makeRow('detail-a2', true, 'parent-a', { hasSubgroupValue: false }),
+          makeRow('parent-b', false, 'none'),
+          makeRow('detail-b1', true, 'parent-b', { hasSubgroupValue: false }),
+        ],
+      });
```
[the remark this answers is not in the document]
Comment (prakharsingh-74) [reply packages/jaeger-ui/src/components/TracePage/TraceStatistics/index.test.jsx:160]: fixed

**Comment by prakharsingh-74** [comment]
@yurishkuro sir, can you please review this PR?

**Comment by prakharsingh-74** [comment]
@parshipcy and @jkowall can you review this PR?

**Comment by yurishkuro** [comment]
Thanks for working on this! The `vitest.config.ts` change is the right fix — narrowing `server.deps.inline` from `true` to a specific allowlist avoids compiling all of `node_modules` through Vite on every run. Please restore the comment explaining why these packages need inlining (CJS/ESM interop under SSR resolve conditions), so future maintainers know how to extend the list if needed.

The test consolidation trades performance for maintainability in a way that isn't necessary. A few specific suggestions:

For `OperationDetailsTable`, the idea of putting all formatting edge cases in a single multi-row dataset and rendering once is sound — but that one render should be done in `beforeAll`, with each behavior kept in its own `it()` block. That way you get the performance benefit of a single mount while preserving failure isolation.

For `TraceStatistics`, the situation is different: most tests mutate component state via `setState`/`togglePopup`, so they need a fresh render per test to avoid ordering dependencies — `beforeAll` doesn't help here. The actual win is in the `searchInTable` tests, which call a pure function on the component instance and don't need a mounted component at all. Extracting those as standalone unit tests of the function itself would reduce renders without any test-quality trade-off.

The 30s timeout is a symptom, not a fix — a test that needs 30 seconds is doing too much work in one go.

**Comment by prakharsingh-74** [comment]
@yurishkuro sir, can you review this PR now?

File: packages/jaeger-ui/src/components/Monitor/ServicesView/operationDetailsTable/index.test.jsx [hunk 5]
```diff
@@ -66,233 +76,134 @@ describe('<OperationTableDetails> with data', () => {
     jest.clearAllMocks();
   });
 
-  it('render No data table', () => {
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/components/Monitor/ServicesView/operationDetailsTable/index.test.jsx:69]: why is there so much structure change in the test?

Comment (prakharsingh-74) [reply packages/jaeger-ui/src/components/Monitor/ServicesView/operationDetailsTable/index.test.jsx:69]: Initially, I consolidated the tests into a single block to speed up Vitest runtimes by avoiding repeated mounts.

However, to restore failure isolation and keep the code clean, I've refactored it to render once in beforeAll while keeping each behavior in its own separate it() block.

@yurishkuro sir

Comment (yurishkuro) [reply packages/jaeger-ui/src/components/Monitor/ServicesView/operationDetailsTable/index.test.jsx:69]: Sorry but I don't have time to go line by line in the diffs. Adding a before-all is a 3 line change, yours is changing and moving almost everything in the test. The change needs to be surgical
